### PR TITLE
Backports 0.15 pr6

### PR DESCRIPTION
**Pull request by PastaPastaPasta**

### CODE REVIEW

File: contrib/devtools/github-merge.py [hunk 1]
```diff
@@ -209,7 +209,7 @@ def main():
             exit(4)
 
         symlink_files = get_symlink_files()
-        for f in symlink_files;
+        for f in symlink_files:
```
Comment (UdjinM6) [inline contrib/devtools/github-merge.py]: This commit (9c8c12e) title should probably say smth like `Merge bitcoin#XXXX: devtools: Fix a syntax error typo`

File: src/txmempool.cpp [hunk 2]
```diff
@@ -359,7 +359,7 @@ CTxMemPool::CTxMemPool(const CFeeRate& _minReasonableRelayFee) :
     // of transactions in the pool
     nCheckFrequency = 0;
 
-    minerPolicyEstimator = new CBlockPolicyEstimator(_minReasonableRelayFee);
+    minerPolicyEstimator = new CBlockPolicyEstimator();
```
Comment (UdjinM6) [inline src/txmempool.cpp]: Should also drop `CTxMemPool::UpdateMinFee()` (no longer used) to make this compilable.

File: contrib/verify-commits/verify-commits.sh [hunk 3]
```diff
@@ -9,26 +9,84 @@
...
+			echo "Tree-SHA512 did not match for commit $1" > /dev/stderr
+			HAVE_FAILED=true
+			return 1
+		fi
+	fi
+
 	local PARENTS
 	PARENTS=$(git show -s --format=format:%P $1)
 	for PARENT in $PARENTS; do
-		if IS_SIGNED $PARENT > /dev/null; then
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply contrib/verify-commits/verify-commits.sh]: Is that not okay? If it works in btc why wouldn't it work here, this was a spot where I just wanted to get closer to upstream

File: src/wallet/wallet.h [hunk 4]
```diff
@@ -9,21 +9,22 @@
 
 #include "amount.h"
 #include "base58.h"
+#include "script/ismine.h"
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/wallet/wallet.h]: Wow, didn't see that one xd

File: contrib/devtools/github-merge.py (hunk 1 above)
Comment (PastaPastaPasta) [reply contrib/devtools/github-merge.py]: original commit was not a merge, but a commit: `309bf1625 devtools: Fix a syntax error typo`

File: contrib/verify-commits/verify-commits.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (UdjinM6) [reply contrib/verify-commits/verify-commits.sh]: I wonder if we are missing some previous PR/commit or are we reverting a change from some future one here?

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -5079,7 +5079,7 @@ void CWallet::postInitProcess(CScheduler& scheduler)
 
     // Run a thread to flush wallet periodically
     if (!CWallet::fFlushScheduled.exchange(true)) {
-        scheduler.scheduleEvery(MaybeCompactWalletDB, 500);
+        scheduler.scheduleEvery(MaybeCompactWalletDB, 500 * 1000);
```
Comment (UdjinM6) [inline src/wallet/wallet.cpp]: Nope, this one is from bitcoin

File: src/txmempool.cpp [hunk 6]
```diff
@@ -1580,13 +1580,6 @@ CFeeRate CTxMemPool::GetMinFee(size_t sizelimit) const {
     return std::max(CFeeRate(rollingMinimumFeeRate), incrementalRelayFee);
 }
 
-void CTxMemPool::UpdateMinFee(const CFeeRate& _minReasonableRelayFee)
```
Comment (UdjinM6) [inline src/txmempool.cpp]: Should also drop it from txmempool.h

File: qa/rpc-tests/p2p-acceptblock.py [hunk 7]
```diff
@@ -224,28 +221,18 @@ def run_test(self):
                     headers_message.headers.append(CBlockHeader(next_block))
                 tips[j] = next_block
 
-        set_mocktime(get_mocktime() + 2)
-        set_node_times(self.nodes, get_mocktime())
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply qa/rpc-tests/p2p-acceptblock.py]: doesn't the added `time.sleep(2)` replace that?

File: src/txmempool.cpp (hunk 6 above)
Comment (PastaPastaPasta) [reply src/txmempool.cpp]: https://github.com/PastaPastaPasta/dash/commit/2ce335ef92fc32f049d18fae976139bbc68bde32

File: src/txmempool.cpp (hunk 2 above)
Comment (PastaPastaPasta) [reply src/txmempool.cpp]: https://github.com/PastaPastaPasta/dash/commit/2ce335ef92fc32f049d18fae976139bbc68bde32

File: qa/rpc-tests/fundrawtransaction.py [hunk 8]
```diff
@@ -187,12 +187,7 @@ def run_test(self):
         dec_tx  = self.nodes[2].decoderawtransaction(rawtx)
         assert_equal(utx['txid'], dec_tx['vin'][0]['txid'])
 
-        try:
-            self.nodes[2].fundrawtransaction(rawtx, {'foo': 'bar'})
-            raise AssertionError("Accepted invalid option foo")
-        except JSONRPCException as e:
-            assert("Unexpected key foo" in e.error['message'])
-
+        assert_raises_jsonrpc(-3, "Unexpected key foo", self.nodes[2].fundrawtransaction, rawtx, {'foo':'bar'})
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply qa/rpc-tests/fundrawtransaction.py]: https://github.com/PastaPastaPasta/dash/commit/2ce335ef92fc32f049d18fae976139bbc68bde32

File: src/wallet/wallet.cpp (hunk 5 above)
Comment (PastaPastaPasta) [reply src/wallet/wallet.cpp]: https://github.com/dashpay/dash/pull/2682/commits/1100b4e26b52eba502539e311cee4fd6a1b1e0c2

File: qa/rpc-tests/p2p-acceptblock.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (UdjinM6) [reply qa/rpc-tests/p2p-acceptblock.py]: No. And adding `time.sleep(2)` is not even a part of the original PR.

[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply qa/rpc-tests/p2p-acceptblock.py]: https://github.com/dashpay/dash/pull/2682/commits/d87a211d53b9d5566ae5b49d6208f1a9ec507f4d

**Comment by PastaPastaPasta** [comment]
Yeah, I see that...
@codablock please review

File: src/wallet/walletdb.cpp [hunk 9]
```diff
@@ -835,129 +835,59 @@ void ThreadFlushWalletDB()
...
-
-    std::unique_ptr<Db> pdbCopy(new Db(dbenv.dbenv, 0));
-    int ret = pdbCopy->open(NULL,               // Txn pointer
-                            filename.c_str(),   // Filename
-                            "main",             // Logical db name
-                            DB_BTREE,           // Database type
-                            DB_CREATE,          // Flags
-                            0);
-    if (ret > 0)
+    if (!IsKeyType(strType) && strType != "hdchain")
```
Comment (codablock) [inline src/wallet/walletdb.cpp]: Hmm, from the original PR it looks like this was a move/refactor of code. The original code removed this line with `hdchain` and re-added it with `hdchain` again. In this PR however, you're removing `hdpubkey` and re-adding `hdchain`, so you're not just moving/refactoring anymore.

I don't know why we have this difference. It is one of the many differences we have in Dash due to the way HD support was merged in back then. Maybe @UdjinM6 knows more.

File: contrib/verify-commits/trusted-keys [hunk 10]
```diff
@@ -1,3 +1,4 @@
 71A3B16735405025D447E8F274810B012346C9A6
-3F1888C6DCA92A6499C4911FDBA1A67379A1A931
+133EAC179436F14A5CF1B794860FEB804E669320
 32EE5C4C3FA15CCADB46ABE529D4BCB6416F53EC
+B8B3F1C0E58C15DB6A81D30C3648A882F4316B9B
```
Comment (codablock) [inline contrib/verify-commits/trusted-keys]: Hmm, these keys are most likely useless for us, as they refer to Bitcoin devs. We should consider using our own keys here in the future.

File: contrib/verify-commits/trusted-sha512-root-commit [hunk 11]
```diff
@@ -0,0 +1 @@
+b00ba6251f71fa1edaabdf809514e1bc3c67862e
```
Comment (codablock) [inline contrib/verify-commits/trusted-sha512-root-commit]: This commit is referring to a Bitcoin commit, so it's useless for us. Should change this to a Dash commit in the future.

File: contrib/verify-commits/verify-commits.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (codablock) [reply contrib/verify-commits/verify-commits.sh]: We're missing https://github.com/bitcoin/bitcoin/pull/9656

File: contrib/verify-commits/trusted-sha512-root-commit [hunk 12]
```diff
@@ -1 +1 @@
-b00ba6251f71fa1edaabdf809514e1bc3c67862e
+f7ec7cfd38b543ba81ac7bed5b77f9a19739460b
```
Comment (codablock) [inline contrib/verify-commits/trusted-sha512-root-commit]: Same as when this file was introduced, refers to a Bitcoin commit

File: src/wallet/walletdb.cpp (hunk 9 above)
Comment (UdjinM6) [reply src/wallet/walletdb.cpp]: Good catch 👍 This should say "hdpubkey" indeed. Original changes https://github.com/dashpay/dash/pull/1405/commits/300e10c8717617e6d812e32609b3582d248c1c2e

Comment (PastaPastaPasta) [reply src/wallet/walletdb.cpp]:
```suggestion
    if (!IsKeyType(strType) && strType != "hdpubkey")
```

**Comment by PastaPastaPasta** [comment]
squashed 2 code review commits

**Pull request merged**